File: python/example4_3.py
# Example 4.3
# Testing for Structural Change

import numpy as np 
import pandas as pd 
from scipy import stats 
import statsmodels.formula.api as smf 

data_cjx=pd.read_csv('http://web.pdx.edu/~crkl/ceR/data/cjx.txt',sep='\s+',nrows=39)
break_year = (data_cjx['YEAR'] > 1948).astype(int)

X = np.log(data_cjx['X']) 
L = np.log(data_cjx['L1'])
K = np.log(data_cjx['K1'])
df = pd.concat([X, L, K], keys=['X', 'L', 'K'], axis=1) 

# restricted model
model1 = smf.ols(formula='X ~ L + K', data=df).fit()
print(model1.summary())

# unrestricted model
model2 = smf.ols(formula='X ~ (L + K)*break_year', data=df).fit()
print(model2.summary())

DFr = model1.df_resid
RSSr = model1.ssr
DFur = model2.df_resid
RSSur = model2.ssr

# F test for structural change
Ftest = ((RSSr-RSSur)/(DFr-DFur))/(RSSur/DFur)
Ftest
# Calculate P-value
1-stats.f.cdf(Ftest, DFr-DFur, DFur)
# threshold = stats.f.ppf(0.95, DFr-DFur, DFur)

# alternatively, do ANOVA
from statsmodels.stats.anova import anova_lm 
anova_lm(model1,model2)

File: R/example4_3.R
# Example 4.3
# Testing for Structural Change

setwd("C:/Course17/ceR/R")
data<-read.table("http://web.pdx.edu/~crkl/ceR/data/cjx.txt",header=T,nrows=39)

year<-data$YEAR
X<-log(data$X)
L<-log(data$L1) 
K<-log(data$K1)

break_year<-factor(year>1948)    # dummy variable

results<-lm(X~(L+K)*break_year)  # unrestricted model
summary(results)
RSSur<-sum(results$residuals^2)
DFur<-results$df.residual

results1<-lm(X~L+K)  # unrestricted model
summary(results1)
RSSr<-sum(results1$residuals^2)
DFr<-results1$df.residual

Ftest<-((RSSr-RSSur)/(DFr-DFur))/(RSSur/DFur)
Ftest
1-pf(Ftest,DFr-DFur,DFur)

# alternatively, do ANOVA
anova(results1,results)

rm(list=ls())
